packages/sdk/src/protocols/filter/reliability_monitor.ts

```diff
@@ -0,0 +1,104 @@
+import type { Peer, PeerId } from "@libp2p/interface";
+import { IProtoMessage, PeerIdStr, PubsubTopic } from "@waku/interfaces";
+import { messageHashStr } from "@waku/message-hash";
+import { WakuMessage } from "@waku/proto";
+import { Logger } from "@waku/utils";
+
+type ReceivedMessageHashes = {
+  all: Set<string>;
+  nodes: Record<PeerIdStr, Set<string>>;
+};
+
+const DEFAULT_MAX_MISSED_MESSAGES_THRESHOLD = 3;
+
+const log = new Logger("sdk:filter:reliability_monitor");
+
+export class ReliabilityMonitor {
+  private receivedMessagesHashes: ReceivedMessageHashes;
+  private missedMessagesByPeer: Map<PeerIdStr, number> = new Map();
+  private maxMissedMessagesThreshold = DEFAULT_MAX_MISSED_MESSAGES_THRESHOLD;
+
+  public constructor(
+    private getPeers: () => Peer[],
+    private renewAndSubscribePeer: (peerId: PeerId) => Promise<Peer | undefined>
+  ) {
+    const allPeerIdStr = this.getPeers().map((p) => p.id.toString());
+
+    this.receivedMessagesHashes = {
+      all: new Set(),
+      nodes: Object.fromEntries(
+        allPeerIdStr.map((peerId) => [peerId, new Set()])
+      )
+    };
+    allPeerIdStr.forEach((peerId) => this.missedMessagesByPeer.set(peerId, 0));
+  }
+
+  public setMaxMissedMessagesThreshold(value: number | undefined): void {
+    if (value !== undefined) {
+      this.maxMissedMessagesThreshold = value;
+    }
+  }
+
+  public addMessage(
+    message: WakuMessage,
+    pubsubTopic: PubsubTopic,
+    peerIdStr?: PeerIdStr
+  ): boolean {
+    const hashedMessageStr = messageHashStr(
+      pubsubTopic,
+      message as IProtoMessage
+    );
+
+    const isNewMessage = !this.receivedMessagesHashes.all.has(hashedMessageStr);
+    this.receivedMessagesHashes.all.add(hashedMessageStr);
+
+    if (peerIdStr) {
+      this.receivedMessagesHashes.nodes[peerIdStr].add(hashedMessageStr);
```

Comment:
it can be that `nodes` still not have `peerId` (e.g after renewal) - we need to check that and initialize a map

Reply:
actually it is the case already, we caught it the other day
![image](https://github.com/user-attachments/assets/7fd42d0d-6217-4b15-ad41-29d118e0e4b7)